### ISSUE TITLE
different inference after converting pt to onnx

### ISSUE THREAD

**Issue by z-habibi** [ultralytics/ultralytics#17465]
### Search before asking

- [ ] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict, Export

### Bug

hello. I have trained a model on yolo and then converted it to onnx.
but the results are not what yolo predicts.
first of all onnx doesn't apply NMS on the predicted anchors. plus the ultralytics predictor cant predict using onnx.

```python
model = YOLO('/content/drive/MyDrive/Colab Notebooks/best.onnx')
results = model("/content/drive/MyDrive/Colab Notebooks/3.jpg")

for result in results:
    boxes = result.boxes
    probs = result.probs
    obb = result.obb
    result.show()
    result.save(filename="result.jpg")`
``` 
but even that not considered, the onnx weights cant predict 
![image](https://github.com/user-attachments/assets/9447edec-95ba-48b3-adf4-161e1710e501)

so I used onnxruntime and used the code below.

```python
session = ort.InferenceSession('/content/drive/MyDrive/Colab Notebooks/plate/best.onnx')
image = cv2.imread(test_image_path)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image = cv2.resize(image, (640,640))
image = np.transpose(image, (2, 0, 1))
image = image.astype(np.float32)
image /= 255.0
image = np.expand_dims(image, axis=0)
input_name = session.get_inputs()[0].name
outputs = session.run(None, {input_name: image})
output = outputs[0]
output_data = output[0]

num_boxes = output_data.shape[1]
box_coords = []
[box_coords.append(output_data[:4,o].tolist()) for o in range(num_boxes)]
class_probs = []
[class_probs.append(output_data[4:,o].tolist()) for o in range(num_boxes)]
score = []
[score.append(max(class_probs[c])) for c in range(len(class_probs))]
classes = np.argmax(class_probs, axis=1)
filter = []
threshold = 0.1
for i in range(len(score)):
  filter.append(score[i]> threshold)

image = cv2.imread(test_image_path)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
plt.imshow(image)
ax = plt.gca()

for i, box in enumerate(box_coords):
  if filter[i]:
    x ,y ,width,height  = box
    rect = plt.Rectangle((x,y) ,width ,height ,linewidth=1.5, edgecolor='blue' , facecolor='none')
    ax.add_patch(rect)
plt.show()
``` 
but the scores are dramatically low.(if I set the `threshold` lower than 0.02 no prediction would remain.this is the output of onnx without any threshold
![image](https://github.com/user-attachments/assets/13ad299c-a28a-4a31-992e-532759c5a3ea)

in the end this is the output of yolo
![image](https://github.com/user-attachments/assets/3eab5b0d-f866-4f57-b774-8b69c7490dc7)

### Environment

Ultralytics 8.3.28 🚀 Python-3.10.12 torch-2.5.0+cu121 CPU (Intel Xeon 2.20GHz)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 32.5/107.7 GB disk)

### Minimal Reproducible Example

session = ort.InferenceSession('best.onnx')
image = cv2.imread(test_image)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image = cv2.resize(image, (640,640))
image = np.transpose(image, (2, 0, 1))
image = image.astype(np.float32)
image /= 255.0
image = np.expand_dims(image, axis=0)
input_name = session.get_inputs()[0].name
outputs = session.run(None, {input_name: image})
output = outputs[0]
output_data = output[0]

num_boxes = output_data.shape[1]
box_coords = []
[box_coords.append(output_data[:4,o].tolist()) for o in range(num_boxes)]
class_probs = []
[class_probs.append(output_data[4:,o].tolist()) for o in range(num_boxes)]
score = []
[score.append(max(class_probs[c])) for c in range(len(class_probs))]
classes = np.argmax(class_probs, axis=1)

image = cv2.imread(test_image)
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
plt.imshow(image)
ax = plt.gca()

for i, box in enumerate(box_coords):
    x ,y ,width,height  = box
    rect = plt.Rectangle((x,y) ,width ,height ,linewidth=1.5, edgecolor='blue' , facecolor='none')
    ax.add_patch(rect)
plt.show()

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
Follow the guide on how to create a [minimum reproducible example ](https://docs.ultralytics.com/help/minimum_reproducible_example/).

What model is this and what variant? What code did you use to export?

**Comment by z-habibi** [comment]
I used ultralytics export

```python
torch_model = YOLO(yolo_weights)
torch_model.export(format='onnx',imgsz=(640,640))
``` 
and by model if you mean ort it's onnxruntime...

**Comment by Y-T-G** [comment]
The model that you exported. (YOLO11n, YOLO11n-seg, etc.)

**Comment by z-habibi** [comment]
Well. the model is a trained yolov8n on my own dataset.

**Comment by z-habibi** [comment]
here are the args of yolo if it helps:
task: detect
mode: train
model: yolov8n.pt
data: D://licpal.yml
epochs: 100
time: null
patience: 100
batch: 16
imgsz: 128
save: true
save_period: 1
cache: false
device: null
workers: 8
project: null
name: train6
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: 4
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 0.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
bgr: 0.0
mosaic: 1.0
mixup: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.4
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml
save_dir: runs\detect\train

**Comment by Y-T-G** [comment]
What size was the model trained at?

Can you post the output of `model.info()` for the PyTorch model?

**Comment by z-habibi** [comment]
Model summary: 225 layers, 3,018,453 parameters, 0 gradients, 8.2 GFLOPs
(225, 3018453, 0, 8.235008)

**Comment by Y-T-G** [comment]
What size was the model trained at?

**Comment by z-habibi** [comment]
what do you mean?

**Comment by Y-T-G** [comment]
What version of `ultralytics` did you use for training?

> Model summary: 225 layers, 3,018,453 parameters, 0 gradients, 8.2 GFLOPs

It's not detecting the model type. It should show the model type:
`YOLOv8n summary: 225 layers, 3,157,200 parameters, 0 gradients, 8.9 GFLOPs`

**Comment by Y-T-G** [comment]
> what do you mean?

`imgsz`

You use `imgsz: 128` for training. Why are you exporting at `640x640`? You need to export at the `imgsz` you used during training.

**Comment by z-habibi** [comment]
okay I exported with imsz 128 and also resized the testing image to 128 but the predictions just got condensed.
![image](https://github.com/user-attachments/assets/6ded1d8b-e360-4760-bcf8-643d1fcdc41b)

**Comment by Y-T-G** [comment]
Is that from ultralytics inference?

**Comment by Y-T-G** [comment]
You should use ultralytics inference. If you want to use ONNXRuntime, then you need to use correct preprocessing and postprocess.

https://github.com/ultralytics/ultralytics/tree/ed252226a9d8d3009ead5a511ab6ba51dafa132a/examples/YOLOv8-ONNXRuntime

**Comment by Y-T-G** [comment]
You should also specify `imgsz=128` when inferring with ultralytics

**Comment by Y-T-G** [comment]
You can't use `result.show()` on Colab. Colab doesn't have a GUI

**Comment by Y-T-G** [comment]
What code did you use? Is that `ultralytics` inference? It doesn't look like it from the annotations. If you're using custom ONNXRuntime, then like I said, you need correct preprocessing and postprocessing.

This example doesn't use letterbox resizing.

https://github.com/ultralytics/ultralytics/tree/ed252226a9d8d3009ead5a511ab6ba51dafa132a/examples/YOLOv8-ONNXRuntime

So it won't be the same as the original.

The correct preprocessing and postprocessing is here.

https://github.com/ultralytics/ultralytics/issues/17423#issuecomment-2466139832

But for ONNX, the input is NCHW while NCNN uses CHW (not batched). You need to add a batch dimension.

**Comment by Y-T-G** [comment]
Use Ultralytics inference and specify `imgsz=128`. 

```pytho
from ultralytics import YOLO

model = YOLO("best.onnx")
results = model("model.jpg", imgsz=128) 
``` 

And the results should be pretty close to PyTorch .

**Comment by z-habibi** [comment]
> Use Ultralytics inference and specify `imgsz=128`.
> 
> ```
> from ultralytics import YOLO
> 
> model = YOLO("best.onnx")
> results = model("model.jpg", imgsz=128) 
> ```
> 
> And the results should be pretty close to PyTorch .

oh this actually worked. thanks!!

**Comment by glenn-jocher** [comment]
Great to hear that it worked for you! If you have any more questions, feel free to ask.
